_posts/2019-01-21-vet-analyzer.md

```diff
@@ -0,0 +1,315 @@
+
+It looks like -
+
+```go
+func run(pass *analysis.Pass) (interface{}, error) {
+}
+```
+
+Now, to run this analyzer, we will use the `singlechecker` package which can be used to run a single analyzer.
+
+```go
+package main
+
+import (
+	"github.com/agnivade/sqlargs"
+	"golang.org/x/tools/go/analysis/singlechecker"
+)
+
+func main() { singlechecker.Main(sqlargs.Analyzer) }
+```
+
+Upon successfully compiling this, you can execute the binary as a standalone tool on your codebase - `sqlargs ./...`.
+
+This is the standard layout of all analyzers. Let us have a look into the internals of the `run` function, which is where the main code analysis is performed.
+
+### Look for SQL queries
+
+Our primary aim is to look for expressions like `db.Exec("<query>")` in the code base and analyze them. This requires knowledge of Go ASTs (Abstract Syntax Tree) to slice and dice the source code and extract the stuff that we need.
+
+To help us with scavenging the codebase and filtering the AST expressions that we need, we have some tools at our disposal, viz. the `go/ast/inspector` package. This package does all the heavy lifting of loading and parsing the source code and just passes on the specified `node` types that we want. Since this is a very common task for all analyzers, we have an `inspect` pass which returns an analyzer that provides an `inspector`.
+
+Let us see how that looks like -
+
+```go
+func run(pass *analysis.Pass) (interface{}, error) {
+	inspect := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	// We filter only function calls.
+	nodeFilter := []ast.Node{
+		(*ast.CallExpr)(nil),
+	}
+
+	inspect.Preorder(nodeFilter, func(n ast.Node) {
+		call := n.(*ast.CallExpr)
+		_ = call // work with the call expression that we have
+	})
+}
+```
+
+All expressions of the form of `db.Exec("<query>")` are called [CallExpr](https://godoc.org/go/ast#CallExpr)s. So we specify that in our `nodeFilter`. After that, the `Preorder` function will give us only `CallExpr`s found in the codebase.
+
+A [CallExpr](https://godoc.org/go/ast#CallExpr) has two parts- Fun and Args. A Fun can either be an [Ident](https://godoc.org/go/ast#Ident) (for eg. `Fun()`) or a [SelectorExpr](https://godoc.org/go/ast#SelectorExpr) (for eg. `foo.Fun()`). Since we are looking for patterns like `db.Exec`, we need to filter only `SelectorExpr`s.
+
+```go
+inspect.Preorder(nodeFilter, func(n ast.Node) {
+	call := n.(*ast.CallExpr)
+	sel, ok := call.Fun.(*ast.SelectorExpr)
+	if !ok {
+		return
+	}
+
+})
+```
+
+Alright, so far so good. This means we have filtered all expressions of the form of `type.Method()` from the source code. Now we need to verify 2 things -
+1. The function name is `Exec`; because that is what we are interested in.
+2. The type of the selector is `sql.DB`.
```

Comment:
I realise at the end you say "nor does the type need to be `*sql.DB`", but I still think the way you've written the check is overly complex. Checking for the method name and type should be enough, and shouldn't cause any false positives.

Reply:
Yeah, it wouldn't catch structs which embed sql.DB. My actual code in the repo does catch it though.

So you mean to say load the type info from `Exec` rather than `db` ?

>  but I still think the way you've written the check is overly complex. Checking for the method name and type should be enough, and shouldn't cause any false positives.

It is not enough because sql queries can come from other libraries like `sqlx` and `gorm`. My actual repo checks for those too. Perhaps I should clarify this further.

Reply:
> So you mean to say load the type info from `Exec` rather than `db` ?

Well, I haven't tested it, but grabbing the `*types.Signature` from the call should be enough to tell if it's the same (or identical) method as the one from the sql package.

Reply:
Thanks. Will check that.

Reply:
Looked into it. The signature does not show the receiver type, nor the method name, which I need. For something like `db.Exec`, it just shows "func(query string, args ...interface{}) (database/sql.Result, error)".